core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -415,17 +415,22 @@ private RexNode simplifyGenericNode(RexCall e) {
    */
   private static int findLiteralIndex(List<RexNode> operands, BigDecimal value) {
     for (int i = 0; i < operands.size(); i++) {
-      if (operands.get(i).isA(SqlKind.LITERAL)) {
-        Comparable comparable = ((RexLiteral) operands.get(i)).getValue();
-        if (comparable instanceof BigDecimal
-            && value.compareTo((BigDecimal) comparable) == 0) {
-          return i;
-        }
+      if (checkLiteralValue(operands.get(i), value)) {
+        return i;
       }
     }
     return -1;
   }
 
+  /** Check whether the operand is a literal of the specified value. */
```

Comment:
This method is a conservative approximation only; the javadoc should say that.
It will not handle for example floating point literals, which can appear due to the constant folding of expressions.

Reply:
I've changed the javadoc to mention BigDecimal.